ASoC: SOF: ipc4/Intel: Support for loading external libraries

ujfalusi opened the pull request:
Hi,

In IPC4 all DSP loadable executable is a 'library' containing modules. The main or basefw is also a library which contains multiple modules.
IPC4 allows to use loadable libraries to extend the functionality of the booted basefw.

This series adds support for loading external libraries in case they are needed by the loaded topology file.

The libraries must be placed to a specific firmware directory (fw_lib_prefix), which is:
intel/avs-lib|sof-ipc4-lib/ followed by the platform name and in case of community key use a 'community' directory.
For example for upx-i11 (community key): `intel/avs-lib/tgl/community` is the default path.

The name of the library should be the UUID of the module it contains since the library loading is going to look for the file as `<module_UUID>.bin`
In case there is a need to bundle multiple modules into single library, symlinks can be used to point to the file:
```
module_boundle.bin
<UUID1>.bin -> module_boundle.bin
<UUID2>.bin -> module_boundle.bin
<UUID3>.bin -> module_boundle.bin
```
But note that in this case all modules will be loaded to the DSP since only the whole library can be loaded, not individual modules.

This PR replaces #3720 from @ranj063 and it is taking completely different route to implement the support, keeping most of the implementation IPC4 specific and not forcing artificial rules to non IPC4 users.

plbossart commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -40,6 +41,7 @@ struct sof_ipc4_fw_data {
 	void *nhlt;
 	enum sof_ipc4_mtrace_type mtrace_type;
 	u32 mtrace_log_bytes;
+	u32 max_fw_libs;
```
naming nit-pick: max_libs_count would be clearer IMHO, with a 1:1 mapping with the query

ujfalusi replied:
OK.
It is unused atm, we might as well want to use it when loading libraries to make sure we are not trying to load more than what is supported.

plbossart commented on sound/soc/sof/loader.c:
```diff
@@ -58,6 +58,9 @@ int snd_sof_load_firmware_raw(struct snd_sof_dev *sdev)
 			fw_filename, ret);
 	}
 
+	/* Platform code is still using the plat_data for firmware handling */
```
it seems that it's a temporary comment, until the platform code is modified? it should be made clearer for reviewers who scan patches one after the other.

ujfalusi replied:
OK.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -444,6 +444,9 @@ struct sof_ipc_pcm_ops;
  * @fw_tracing:	Pointer to Firmware tracing ops
  *
+ * @init:	Optional pointer for IPC related initialization
+ * @fini:	Optional pointer for IPC related cleanup
```
I personally have have no idea what 'fini' would stand for. We have no precedent for using 'fini'. can we find a more usual name for this callback? 

If we use the ASoC convention, we have init/exit.

ujfalusi replied:
Yeah, it is coming from DRM, but sure, let's use `exit`

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -122,6 +122,35 @@ static size_t sof_ipc4_fw_parse_ext_man(struct snd_sof_dev *sdev)
+	payload_offset = sof_ipc4_fw_parse_ext_man(sdev, fw_lib);
+	if (payload_offset > 0) {
+		fw_lib->sof_fw.payload_offset = payload_offset;
+		id = idr_alloc(&ipc4_data->fw_lib_idr, fw_lib, 0, 0, GFP_KERNEL);
```
In the kernel documentation https://www.kernel.org/doc/html/latest/core-api/idr.html#:~:text=The%20IDR%20provides%20the%20ability,is%20much%20more%20memory%2Defficient. I see this:

The IDR interface is deprecated; please use the XArray instead.

??

ujfalusi replied:
Oh, that's unfortunate. The IDR API is straight forward with lots of users... I'll study XArray, it would be a shame if I would need to do this by hand (list+bitfiled+lookups).

plbossart commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -24,40 +24,55 @@ enum sof_ipc4_mtrace_type {
 
+/**
+ * struct sof_ipc4_fw_module - IPC4 module info
+ * @sof_man4_module : Module info
```
spurious space after _module

ujfalusi replied:
True.

plbossart commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -24,40 +24,55 @@ enum sof_ipc4_mtrace_type {
+ */
+struct sof_ipc4_fw_module {
+	struct sof_man4_module man4_module_entry;
+	struct ida m_ida;
```
this is very confusing, the m_ida is clearly related to a module instance but the bss_size would be common to all instances of that module, no? 

Or is the term 'module instance' an oversight, this is just the ID for a specific module that could be instantiated multiple times, e.g. on different pipelines?

ujfalusi replied:
Yes, the later. One module which can be instantiated multiple times.
Note: I just moved the struct and comment w/o modification up...

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -151,6 +151,23 @@ static size_t sof_ipc4_fw_parse_basefw_ext_man(struct snd_sof_dev *sdev)
+	struct sof_ipc4_fw_library *fw_lib;
+	int i, id;
+
+	idr_for_each_entry(&ipc4_data->fw_lib_idr, fw_lib, id) {
```
s/id/lib_id/  for clarity? I really hate the mixed use of i and id...

ujfalusi replied:
OK, I can do that in similar places.

plbossart commented on sound/soc/sof/sof-pci-dev.c:
```diff
@@ -272,6 +276,27 @@ int sof_pci_probe(struct pci_dev *pci, const struct pci_device_id *pci_id)
+					sof_pdata->desc->default_lib_path[sof_pdata->ipc_type],
+					"community");
+
+			dev_dbg(dev, "Platform uses community key, changed fw path to %s\n",
```
copy-paste,  should be "changed lib_path to"

ujfalusi replied:
Oh, yes. Thanks!

plbossart commented on sound/soc/sof/intel/pci-apl.c:
```diff
@@ -33,6 +33,9 @@ static const struct sof_dev_desc bxt_desc = {
 		[SOF_INTEL_IPC4] = "intel/avs/apl",
 	},
+	.default_lib_path = {
+		[SOF_INTEL_IPC4] = "intel/sof-lib/apl",
```
we use intel/sof-ipc4/apl for the base firmware, should there be a mention of ipc4 for the libs as well?

If there is ever support for libraries for IPC3 the suggested naming would become ambiguous.

intel/sof-ipc4-lib ?

ujfalusi replied:
Actually it is `intel/avs/apl`, only MTL have `intel/sof-ipc4/mtl`
Should we have `avs-lib` and `sof-ipc4-lib` depending on where the IPC4 default path is pointing?

plbossart replied:
probably yes, that would be clearer

plbossart commented on sound/soc/sof/intel/pci-skl.c:
```diff
@@ -52,6 +55,9 @@ static struct sof_dev_desc kbl_desc = {
 	},
+	.default_lib_path = {
+		[SOF_INTEL_IPC4] = "intel/sof-lib/kbl",
+	},
```
wondering if the library loading is the same for SKL/KBL, it's not the same loader at all. Maybe we should put this on the back-burner for now, it's not clear how we would test this and it's not really a priority, is it?

ujfalusi replied:
OK, I'll drop this and the setting of the load_library callback for SKL.

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -396,12 +396,16 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
 		ret = hda_dsp_boot_imr(sdev);
-		if (!ret)
+		if (!ret) {
+			hda->booted_from_imr = true;
```
this is surprising, we already skip the base firmware loading when booting from IMR today, not sure why we need the same information stored again for libraries?

ujfalusi replied:
I don't want to re-use the same flag for the decision to **do** or **not do** an IMR boot vs if we **booted** or **not booted** from IMR.
IMR boot can fail and then we will try a cold boot. I'm certain that we will bump into issues when re-purposing signals.

plbossart replied:
that's fine, maybe the definition needs to be a bit clearer so that it's clear that it's just a record of what happened during boot, with the decision depending on other criteria. I clearly missed the nuance when reading the code.

ujfalusi replied:
The commit message in https://github.com/thesofproject/linux/pull/3826/commits/b49f6c4ca971efa973d31fd9162c70b7b03df8c8 should be expanded?
Or shall we have comment? I think I have it when checking the flag, but not when setting it.

plbossart replied:
yes, commit message is fine.

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -518,6 +520,70 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
+	int ret, ret1;
+
+	/* IMR booting will restore the libraries as well, skip the loading */
+	if (reload && hda->booted_from_imr)
```
don't we need to reload if the firmware crashed? we've done all this logic for the base firmware, not sure what's different about the additional libraries.

We also know that IMR is not preserved when hibernating, so that flags needs to be ignored in at least two cases (crash and hibernate).

ujfalusi replied:
We will reload since the `booted_from_imr` is going to be false, so we continue to load the library.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -425,12 +425,15 @@ struct sof_ipc_pm_ops {
  *			configuration from the booted firmware.
  *			Executed after the first successful firmware boot.
+ * @reload_fw_addons:	Optional function pointer to reload firmware components
+ *			after the base firmware has been booted up
```
naming nit-pick: why are we talking about 'reloading", that would mean we've loaded those component already. Did you mean 'load_fw_addons'?

and I also don't get why we can't be explicit about libraries, I don't know what 'addons' might mean, that could mean very different things to different people.

And we also need to be clear that there is no 'unload'

ujfalusi replied:
Right, my idea with this was to give a point where IPC dependent post firmware boot loading, patching can be executed. Be it libraries, some blobs, tuning values, whatever.
Basically a fixed point right after firmware boot to download (if needed) anything that the firmware might need prior to continuing to configure and run.

I have chosen reload, but perhaps `restore` might be better? If it is load_fw_addons then this callback must be also called after the first boot, right? Without condition, after `query_fw_configuration()`

kv2019i replied:
@ujfalusi Isn't this quite close to existing "post_fw_run()" at DSP ops level? Maybe just have same for ipc_fw_loader ops (i.e. define post_fw_run also for fw_loaded_ops...?

ujfalusi replied:
@kv2019i, @plbossart, yes, in a way it is pretty close to the platform's `post_fw_run()` with a distinct difference:
the `post_fw_run()` is for IPC agnostic, hardware, platform related things to do after the fw has booted.

One way we could do this I think is to:
Add generic, IPC level callback: `post_fw_run()` to `struct sof_ipc_ops`
Call that after the platform's post_fw_run() without condition (if set)
In IPC4, if it is the first run, do what we do in `query_fw_configuration()` otherwise reload libraries
remove the `query_fw_configuration()` from sof_ipc_fw_loader_ops and from IPC4.

This will somehow makes things better balanced and generic enough to be usable for purposes we are not yet seeing.

ujfalusi replied:
right, so it is really need to be in the fw_loader_ops. I'll do that.

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -151,12 +165,81 @@ static size_t sof_ipc4_fw_parse_basefw_ext_man(struct snd_sof_dev *sdev)
+
+	/* Fix up the module ID numbers within the library */
+	for (i = 0; i < fw_lib->num_modules; i++)
+		fw_lib->modules[i].man4_module_entry.id += (SOF_IPC4_MODULE_ID_STRIDE * id);
```
I could not follow what this is doing? Is this a firmware requirement, or some of sort of trick?

ujfalusi replied:
Modules must have unique id numbers. Afaik there is no hard definition on the stride, 0x1000 is just a safe distance. Let me check, but it might be required by the protocol.

ujfalusi replied:
The same stride is used within the firmware for modules within a libraries.

plbossart reviewed (changes requested):
Looks mostly good @ujfalusi, I added a number of nit-picks comments. My main concern is the IMR boot, this doesn't seem to be aligned with fixes we added for the base firmware for crashes and hibernate.

ujfalusi commented:
> Looks mostly good @ujfalusi, I added a number of nit-picks comments. My main concern is the IMR boot, this doesn't seem to be aligned with fixes we added for the base firmware for crashes and hibernate.

The new flag `booted_from_imr` will going to take care of that. I have decided to not play with existing flags decisions dealing with deciding if we should boot from IMR, but with the new flag we _know_ if the firmware was booted up from IMR or not.
This will take care of the cases when we force the cold DSP boot after either firmware crash or coming up from hibernation.

plbossart commented:
> > Looks mostly good @ujfalusi, I added a number of nit-picks comments. My main concern is the IMR boot, this doesn't seem to be aligned with fixes we added for the base firmware for crashes and hibernate.
> 
> The new flag `booted_from_imr` will going to take care of that. I have decided to not play with existing flags decisions dealing with deciding if we should boot from IMR, but with the new flag we _know_ if the firmware was booted up from IMR or not. This will take care of the cases when we force the cold DSP boot after either firmware crash or coming up from hibernation.

I clearly missed where this flag is reset based on higher level information (crash or hibernate).

ujfalusi commented:
> > > Looks mostly good @ujfalusi, I added a number of nit-picks comments. My main concern is the IMR boot, this doesn't seem to be aligned with fixes we added for the base firmware for crashes and hibernate.
> > 
> > 
> > The new flag `booted_from_imr` will going to take care of that. I have decided to not play with existing flags decisions dealing with deciding if we should boot from IMR, but with the new flag we _know_ if the firmware was booted up from IMR or not. This will take care of the cases when we force the cold DSP boot after either firmware crash or coming up from hibernation.
> 
> I clearly missed where this flag is reset based on higher level information (crash or hibernate).

It is never cleared as such. It is _set_ to indicate exactly how the DSP got booted up. The decision to skip IMR should not be of concern for the library loading, what matters is how the DSP booted up. We might tried to boot from IMR, but failed to do that and done a clean boot instead.

kv2019i reviewed (commented):
Good work @ujfalusi and @ranj063 ! The reworked IPC abstraction is good, this makes it easier to support another library load mechanism on the side. The series seems very clean to me and easy to read the commit messages. Deprecation of IDR seems unfortunate, that helped code readability as well. Let's see the XArray version. One minor comment inline.

ujfalusi commented:
Changes since v1:
- Use XArray instead of the deprecated IDR for managing libraries
- The path for the libraries are now aligned with the firmware path (avs-lib or sof-ipc4-lib, depending on the platform)
- max_fw_libs renamed to max_libs_count and it is now used when allocating IDs in XArray
- The query_fw_configuration fw_loader ops is removed and replaced by generic ipc_ops.post_fw_boot ops
- Commit message expanded for the new booted_from_imr flag
- `fini` is now `exit` for IPC level cleanup function
- `lib_id` is used in place of just `id`
- copy paste comments  and typos fixed

ujfalusi commented:
SOFCI TEST

plbossart commented:
SOFCI TEST

RanderWang commented on sound/soc/sof/intel/hda.h:
```diff
@@ -856,4 +857,7 @@ irqreturn_t hda_dsp_ipc4_irq_thread(int irq, void *context);
 int hda_dsp_ipc4_send_msg(struct snd_sof_dev *sdev, struct snd_sof_ipc_msg *msg);
 extern struct sdw_intel_ops sdw_callback;
 
+struct sof_ipc4_fw_library;
```
?  what's this ?

ujfalusi replied:
forward declaration so we don't need to include the whole `ipc4-priv.h` to `hda.h`.

RanderWang replied:
advance usage

RanderWang commented on sound/soc/sof/ipc4.c:
```diff
@@ -656,7 +657,44 @@ static const struct sof_ipc_pm_ops ipc4_pm_ops = {
+
+	xa_for_each(&ipc4_data->fw_lib_xa, lib_id, fw_lib) {
+		/* Do not release the basefw library */
+		if (lib_id != 0)
```
if (lib_id) ?

ujfalusi replied:
Normally I would prefer `if (lib_id)` but here I want to be explicit about the fact that we are excluding the basefw, which is always ID0.

RanderWang reviewed (commented):
LGTM. Do we plan to support unload lib ?

ujfalusi commented:
> LGTM. Do we plan to support unload lib ?

At this point it is not planned. When you would need such feature you would need to load a new tplg file, so reboot or module removals.
No plan for on demand loading either at this point to load only libraries needed by the started pipeline, we load all needed libraries by the tplg all the time

ujfalusi commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -167,13 +170,86 @@ static size_t sof_ipc4_fw_parse_basefw_ext_man(struct snd_sof_dev *sdev)
+
+	kfree(fw_filename);
+
+	return ipc4_data->load_library(sdev, fw_lib, false);
```
Note: we will keep the library in the kernel list even if loading fails. But this will only happen during tplg parsing at the moment and if that fails, the probe will also fail and thus it does not really matter since the SOF stack is not initialized to be usable.

Other use cases will need more planning to cover all corner cases.

ujfalusi replied:
I will not leave it like this. Let me find a solution which has no holes and TODOs.

ujfalusi commented:
Changes since v2:
- The first attempt pf loading a new library is done while making sure that the DSP is booted
- Remove fw_lib from kernel list in case the actual loading of it to DSP fails
- free up all memory if library add/load fails

This change will allow loading libraries not requested by tplg file.

ujfalusi commented:
I think it is still not 100% correct: if IMR boot is _not_ enabled/possible and we try to load a library when the DSP was off:
we boot up the DSP, the post_fw_boot will load the libraries in the XArray and this list includes the new lib that we are actually trying to load.
The XArray allocation needs to be the last thing done in the `sof_ipc4_load_library_by_uuid()` to fix this.

After that I think it should be ready.

plbossart commented:
> This change will allow loading libraries not requested by tplg file.

@ujfalusi how would the kernel know about such libraries? you've said too much or too little here.... Do we want to hard-code UUIDs in the kernel? provide a debugfs path or something?

ujfalusi commented:
> > This change will allow loading libraries not requested by tplg file.
> 
> @ujfalusi how would the kernel know about such libraries? you've said too much or too little here.... Do we want to hard-code UUIDs in the kernel? provide a debugfs path or something?

The UUIDs are tied to specific functionality, for example GDB support can be loaded as a library.
If we will have support for such functionality then the code dealing with it can use the `sof_ipc4_find_module_by_uuid()` call to check if the support is availabe (if not then try to load it).

ujfalusi commented:
Changes since v3:
- Use xa_insert() to a specific index in the XArray
- only add the new library to the XArray when the loading of it was successful

This version should work under all scenarios I can think of.

ujfalusi commented:
@plbossart, we _could_ provide a debugfs file to force load a given UUID (or check it's existence) but I would be cautious about it as we don't have a way to remove them.
Hardcoding UUID should be OK for specific features needing driver support (GDB for example). For non autonomous modules? Probably debugfs file, but at this point I don't have plan for that.

RanderWang commented:
> @plbossart, we _could_ provide a debugfs file to force load a given UUID (or check it's existence) but I would be cautious about it as we don't have a way to remove them. Hardcoding UUID should be OK for specific features needing driver support (GDB for example). For non autonomous modules? Probably debugfs file, but at this point I don't have plan for that.

why GDB module can't be a built-in one in basefw ? We can disabled or enable it by kconfig

ujfalusi commented:
> > This change will allow loading libraries not requested by tplg file.
> 
> @ujfalusi how would the kernel know about such libraries? you've said too much or too little here.... Do we want to hard-code UUIDs in the kernel? provide a debugfs path or something?

We have for example GDB support as library (UUID: 5842A646-15AC-4CA4-BD0B-53D5D40822BB). If we add support for using GDB for debugging then the driver which handles this would need to request this specific UUID to load the external GDB lib to the firmware. This UUID is constant and unique.

This is one example.

plbossart commented:
@ujfalusi same question as @RanderWang, the GDB stub part doesn't seem like an extension or something that should be handled by 3rd parties.

ujfalusi commented:
> why GDB module can't be a built-in one in basefw ? We can disabled or enable it by kconfig

I guess it can be, but it does not really make much difference for the driver code which handles it. It will look for the UUID specific to the GDB support, if it is already in basefw, then fine, it proceeds to do whatever needs to be done. If it is not in the basefw then we will look for it in the filesystem as library and load it, then the code can do the same thing as it did when the GDB was built in. It is transparent for the code: either the module with a given UUID is present or not, it does not matter for the GDB support code that it was built in or loaded as library.

ujfalusi commented:
We could have _everything_ built as module, copier, dmic, ssp, gain. It will not matter for the kernel as they will be loaded to firmware. I'm not saying that it makes sense for these basic and mandatory components, I'm just saying that this is really something that the rest of the code should not know about.

plbossart commented:
@ujfalusi but it matters in terms of releases. if the GDB stuff is not included in the base firmware, then Intel will have to release it separately in sof-libs, and add a hard-coded reference in the kernel. Why would we create extra work for ourselves?

ujfalusi commented:
> @ujfalusi but it matters in terms of releases. if the GDB stuff is not included in the base firmware, then Intel will have to release it separately in sof-libs, and add a hard-coded reference in the kernel. Why would we create extra work for ourselves?

Intel does not _need_ to release it, but it is not changing the fact that for example GDB have a defined UUID assigned which is not changing.
The gdb support code can either check if the given UUID is in the firmware (and thus try to load it as lib if it is not included) or try some IPC and take the failure as a no, the GDB support is not present and gave up.

What I was trying to say is that now the code is going to be able to handle things that might be needed, we do not need to rework it again and the core features are not affected by this forward looking implementation.

If I mention GDB in any of the commit message then I will remove it. The PR is to handle external library loading when the given module with a requested UUID is not in the loaded firmware.

RanderWang commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -178,6 +281,30 @@ struct sof_ipc4_fw_module *sof_ipc4_find_module_by_uuid(struct snd_sof_dev *sdev
+	if (ret)
+		return NULL;
+
+	/* Look for the module in the newly loaded library, it should be available now */
```
the fw_lib is built by sof_ipc4_load_library_by_uuid, so we can return it by this function. Why we need to find it again ?

ujfalusi replied:
The `sof_ipc4_find_module_by_uuid()` returns a pointer to the **fw module**, I think it would be awkward for the `sof_ipc4_load_library_by_uuid()`, which is loading a library (which can contain multiple modules) to return a pointer to a module also.
The very same loop must be done anyways as the manifest parsing is _not_ identifying the modules.

I think keeping the layers separate is better.

ujfalusi commented:
Changes since v4:
- Use shift for setting the lib_id within the module_id (lib_id starts at bit 12)

RanderWang reviewed (approved):
LGTM

ranj063 commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -24,38 +24,53 @@ enum sof_ipc4_mtrace_type {
  * @manifest_fw_hdr_offset: FW header offset in the manifest
- * @num_fw_modules : Number of modules in base FW
- * @fw_modules: Array of base FW modules
+ * @fw_lib_idr: IDR for firmware libraries, including basefw (ID = 0)
```
fw_lib_xa?

ujfalusi replied:
oh, true.

ranj063 commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -178,6 +282,30 @@ struct sof_ipc4_fw_module *sof_ipc4_find_module_by_uuid(struct snd_sof_dev *sdev
+		return NULL;
+	}
+
+	/* The module can not be found, try to load it as a library */
```
we should move this comment before line 278

ujfalusi replied:
Why? In the first loop we are checking the already loaded libraries. At this very point we start to look for the library from the filesystem as it was not found.

ranj063 commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -178,6 +282,30 @@ struct sof_ipc4_fw_module *sof_ipc4_find_module_by_uuid(struct snd_sof_dev *sdev
+			if (guid_equal(uuid, &fw_lib->modules[i].man4_module_entry.uuid))
+				return &fw_lib->modules[i];
+		}
+	}
```
can we condense the function to do this search only once by calling sof_ipc4_load_library_by_uuid() unconditionally first ? If the library is already loaded, this function should pretty much do nothing

ujfalusi replied:
But then we should have similar check in the load library function to check if the requested library has been already loaded.
The first loop will check the loaded libraries, if the module is found it returns. If not then loads a library from the filesystem, parses the ext_manifest, then we will start to look for the new module starting _after_ the last library we have checked in the first loop, so we will continue the loop with the just loaded lib.

I don't want to put extra logic to load llibrary to deal with finding and matching modules, it is not it's job.

ranj063 commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -270,6 +398,32 @@ int sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
+
+	/*
+	 * It is not an error if the callback is not set at this point since the
+	 * loaded topology did not needed external libraries to be loaded
```
typo: did not need

ranj063 replied:
not sure I follow this comment @ujfalusi. Why is not an error "at this point"?

ujfalusi replied:
We call this function unconditionally on non first boot cases. If the ipc4_data->load_library is not set and we don't have external libraries then it is just fine to return with 0.
If we would have needed external libraries and the ipc4_data->load_library is not set then the tplg parsing would have failed.

I don't think the upper layer should know much about libraries, number of libraries, etc.

The check is for sanity and might be dropped since the xa_for_each_start() would have 0 run in this case...

ujfalusi replied:
I think it is safe to drop this check and the comment (along with the typo)

ranj063 reviewed (commented):
@ujfalusi this is pretty impressive. just a couple of comments

ujfalusi commented:
Changes since v5:
- fixed comment documentation of fw_lib_xa
- Removed checking of `ipc4_data->load_library` and the comment in reload.

kv2019i reviewed (approved):
Refreshing my approval.

ranj063 commented on sound/soc/sof/ipc.c:
```diff
@@ -217,5 +220,8 @@ void snd_sof_ipc_free(struct snd_sof_dev *sdev)
 	ipc->disable_ipc_tx = true;
 	mutex_unlock(&ipc->tx_mutex);
+
+	if (ipc->ops->exit)
```
why don't you check is ipc->opc->exit is set here?

ujfalusi replied:
Hrm, I do check it, no?

ranj063 commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -178,6 +282,30 @@ struct sof_ipc4_fw_module *sof_ipc4_find_module_by_uuid(struct snd_sof_dev *sdev
+		return NULL;
+	}
+
+	/* The module can not be found, try to load it as a library */
```
typo can not is one word

ujfalusi replied:
Thanks

ranj063 commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -160,13 +164,113 @@ static size_t sof_ipc4_fw_parse_basefw_ext_man(struct snd_sof_dev *sdev)
+	}
+
+	fw_lib->sof_fw.payload_offset = payload_offset;
+	fw_lib->id = lib_id;
```
@ujfalusi i think there's a library offset of 0x1000 for external libraries @aiChaoSONG correct me if I am wrong please

aiChaoSONG replied:
the offset is not for library id, it is for module id in the loadable library.
The module id in a loadable lib is calculated like this: `*module_id = (library_idx << 12) + local_mod_id;`

ranj063 replied:
@ujfalusi last question, are we handling this correctly in this PR?

ujfalusi replied:
> @ujfalusi last question, are we handling this correctly in this PR?

The library id is a running integer number.
The module's ID is a composite of library ID (which the module belongs to) and a running integer for the module ID.

To not mess around with the generic ext manifest parsing where we assign IDs to the modules within the library I have opted to update the module ID here at line 221.

ranj063 replied:
ok that makes me wonder line 221 is the < correct?

ujfalusi replied:
@ranj063, you have good eye ;) It should have been `<<` as shift..

ujfalusi commented:
Changes since v6:
- type fixed: /can not/cannot

This unfortunately resets again the approvals :(

jsarha reviewed (dismissed):
Not suggesting the currently requested changes are not needed, not even claiming that I fully understand what this PR is all about. However, I browsed the PR through, could not find anything seriously wrong, tested my IPC4 probes changes, using sof_ipc4_find_module_by_uuid(),  on top of this, and it still worked.

jsarha commented:
> Not suggesting the currently requested changes are not needed, not even claiming that I
> fully understand what this PR is all about. However, I browsed the PR through, could not
> find anything seriously wrong, tested my IPC4 probes changes, using
> sof_ipc4_find_module_by_uuid(), on top of this, and it still worked.

Oops, my code worked on top of the previous version (5c8684c59a8fbd663e21db9068af721b8a524ac1), but with this latest (84e58e96f4f04f1036b770e6e6254f011e9abd23) I can not even get playback started.

plbossart commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -24,38 +24,53 @@ enum sof_ipc4_mtrace_type {
+ */
+struct sof_ipc4_fw_library {
+	struct sof_firmware sof_fw;
+	u32 id;
```
lib_id or library_id then?

ujfalusi replied:
I think that the `sof_ipc4_fw_library.id` is indicating that this is the id of the library and when used as `fw_lib->id` is also clear on what the id might be.
`fw_lib->lib_id` is a bit too much `lib` for me.

plbossart commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -24,38 +24,53 @@ enum sof_ipc4_mtrace_type {
+ * struct sof_ipc4_fw_library - IPC4 library information
+ * @sof_fw: SOF Firmware of the library
+ * @id: Library ID. 0 is reserved for basefw, external libraries must have unique
+ *	ID number starting from 1
```
is there a maximum?
is this actually starting from 1 with contiguous values or can the library ID be non-contiguous with a value > 0?

ujfalusi replied:
Yes, there is a limit, which is going to be queried from the firmware and saved in `ipc4_data->max_libs_count`

It is not a requirement that they are to be contiguous.

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -657,7 +658,35 @@ static const struct sof_ipc_pm_ops ipc4_pm_ops = {
+	unsigned long lib_id;
+
+	xa_for_each(&ipc4_data->fw_lib_xa, lib_id, fw_lib) {
+		/* Do not release the basefw library */
```
maybe add a comment on when the basefw library is released?

I actually don't understand why we special case the basefw? In the commit message you stated "This patch abstracts away the difference between basefw and other libraries
to make it possible to implement external module loading." but this code does add a difference.

ujfalusi replied:
OK, I'll update the comment.

We special case the basefw because it is a special case. The basefw is the kernel for the DSP and it is loaded by the generic loader code. The basefw is the kernel for the DSP, there is no difference on how the kernel is loaded by the SOF stack and it is handled by a generic code.

For IPC4 we just 'borrow' what has been already loaded as Library #0 to be able to handle generic code to walk the libraries later.
This patch do abstracts the difference between basefw and external libraries by representing the already present basefw as a library.
We _could_ move this whole library concept up to the highest level and force it to IPC3, which does not have support for external libraries, afaik.
Moving this up would create another question: is this library and modules inside of a library concept universal? HOw do we abstract away this detail? Do we represent firmware files and then leave the interpretation to IPC code? How the information will flow from IPC to generic and backwards? Is it written in stone that UUID is the name that needs to be looked up by other IPC types? So on.

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -160,13 +164,113 @@ static size_t sof_ipc4_fw_parse_basefw_ext_man(struct snd_sof_dev *sdev)
+
+	/* Fix up the module ID numbers within the library */
+	for (i = 0; i < fw_lib->num_modules; i++)
+		fw_lib->modules[i].man4_module_entry.id |= (lib_id < SOF_IPC4_MOD_LIB_ID_SHIFT);
```
that convoluted formula is worthy of a comment, not sure why the module_entry.id would be ORed with a boolean?

ujfalusi replied:
Sorry, the lib_id should be shifted up, so it is:
```c
fw_lib->modules[i].man4_module_entry.id |= (lib_id << SOF_IPC4_MOD_LIB_ID_SHIFT);
```

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -160,13 +164,113 @@ static size_t sof_ipc4_fw_parse_basefw_ext_man(struct snd_sof_dev *sdev)
+	kfree(fw_filename);
+free_fw_lib:
+	devm_kfree(sdev->dev, fw_lib->modules);
+	devm_kfree(sdev->dev, fw_lib);
```
what's the point of using devm_ if we used devm_kfree here? why not use plan kzalloc/kfree?

ujfalusi replied:
It makes the normal path cleaner by not requiring brain work on how to safely clean up in `sof_ipc4_exit()`

plbossart reviewed (changes requested):
mostly nit-pick comments, this PR isn't always easy to follow.

ujfalusi commented:
Changes since v7:
- the maximum number of libraries supported by the firmware reading move earlier
- commit message for the the introduction of firmware library abstraction updated
- comments extended where it was requested
- Fixed the lib_id shifting in place within the module_id.

jsarha commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -160,13 +164,113 @@ static size_t sof_ipc4_fw_parse_basefw_ext_man(struct snd_sof_dev *sdev)
+	int i, ret;
+
+	if (guid_is_null(uuid))
+		return NULL;
```
The guid_is_null() check has somehow doubled here.

ujfalusi replied:
Oh, I really wanted to make sure that it is not NULL ;)

ujfalusi commented:
Changes since v8:
- duplicated uuid_is_null() check removed, caused by some rebase and patch shuffling.

ranj063 commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -160,13 +164,110 @@ static size_t sof_ipc4_fw_parse_basefw_ext_man(struct snd_sof_dev *sdev)
+	ret = request_firmware(&fw_lib->sof_fw.fw, fw_filename, sdev->dev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "Library file '%s' is missing\n", fw_filename);
+		goto free_filename;
```
@ujfalusi last question from me (hopefully). If request_firmware fails?, we would go to free_filename which would free fw_lib->modules and fw_lib. But fw_lib->modules is likely still NULL right? I suppose its OK but just checking

ujfalusi replied:
devm_kfree() handles the NULL, so it is not an issue, but you are right that the fw_lib->modules should have been at the release: label..

ujfalusi commented:
Changes since v9:
- Move the devm_kfree(sdev->dev, fw_lib->modules); up in `sof_ipc4_load_library_by_uuid()` and add a comment as well.

aiChaoSONG commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -202,6 +202,13 @@ static int sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
+		case SOF_IPC4_FW_CFG_MAX_LIBS_COUNT:
+			trace_sof_ipc4_fw_config(sdev, "maximum number of libraries",
+						 *tuple->value);
+			ipc4_data->max_libs_count = *tuple->value;
+			if (!ipc4_data->max_libs_count)
+				ipc4_data->max_libs_count = 1;
```
I think `ipc4_data->max_libs_count = *tuple->value ? *tuple->value : 1;` is simpler, but anyway, the existing one is already very readable.

ujfalusi replied:
@aiChaoSONG, I rather not touch this PR now to loose the Approvals, maybe a fixup if this get merged can address this?

aiChaoSONG replied:
sure

aiChaoSONG reviewed (approved):
I have tested the initial draft from Ranjani, at that time, the module loaded at runtime seems cannot work after dsp suspended. Module reloading is handled in the PR, so I will suppose it will work. Thanks.

ujfalusi commented:
> I have tested the initial draft from Ranjani maybe several months ago, at that time, the module loaded at runtime seems cannot work after dsp resuming from suspend. Module reloading is handled in the PR, so I will suppose it will work. Thanks.

@aiChaoSONG, for the library availability after suspend: if IMR boot is used then the library/module is reloaded via that method, if the IMR boot is disabled or not supported on the platform then with this PR we are going to re-load them in the same order as they were loaded initially.
